Aplib.Tests/Core/Belief/BeliefTests.cs

```diff
@@ -1,87 +1,95 @@
 ﻿using Aplib.Core.Belief;
+using System.Linq;
 
 namespace Aplib.Tests.Core.Belief;
+
+/// <summary>
+/// Describes a set of tests for the <see cref="Belief{TReference,TObservation}"/> class.
+/// </summary>
 public class BeliefTests
 {
-    private static TReference ID<TReference>(TReference reference) => reference;
-
-    private static int GetCount<T>(IEnumerable<T> reference) => reference.Count();
-
-    private static bool TrueUpdateIf() => true;
-
-    private static bool FalseUpdateIf() => false;
+    /// <summary>
+    /// A constant 'true' method for testing.
+    /// </summary>
+    /// <returns>True.</returns>
+    private static bool AlwaysUpdate() => true;
 
+    /// <summary>
+    /// A constant 'false' method for testing.
+    /// </summary>
+    /// <returns>False.</returns>
+    private static bool NeverUpdate() => false;
 
     /// <summary>
     /// Given a Belief instance,
-    /// When it is assigned to a variable of its resource type,
-    /// Then it is implicitly converted to its resource type.
+    /// When it is assigned to a variable of its observation type,
+    /// Then it is implicitly converted to its observation type.
     /// </summary>
     [Fact]
-    public void Belief_AssignedToResourceType_IsImplicitlyConvertedToResourceType()
+    public void Belief_AssignedToObservationType_IsCorrectlyImplicitlyConvertedToObservationType()
     {
         // Arrange
-        string reference = "def";
-        Belief<string, string> belief = new(reference, ID);
+        string def = "def";
+        Belief<string, string> belief = new(def, reference => reference);
 
         // Act
-        string resource = belief;
+        string observation = belief;
 
         // Assert
-        Assert.Equal("def", resource);
+        Assert.Equal(def, observation);
     }
 
     /// <summary>
-    /// Given a Belief instance with an updateIf condition that is satisfied,
+    /// Given a Belief instance with an shouldUpdate condition that is satisfied,
     /// When UpdateBelief is called,
-    /// Then the resource is updated.
+    /// Then the observation is updated.
     /// </summary>
     [Fact]
-    public void UpdateBelief_UpdateIfConditionIsSatisfied_UpdatesResource()
+    public void UpdateBelief_ShouldUpdateConditionIsSatisfied_UpdatesObservation()
     {
         // Arrange
         List<int> list = [];
-        Belief<List<int>, int> belief = new(list, GetCount, TrueUpdateIf);
+        Belief<List<int>, int> belief = new(list, reference => reference.Count, AlwaysUpdate);
 
         // Act
         list.Add(69);
         belief.UpdateBelief();
 
         // Assert
-        Assert.Equal(1, belief);
+        Assert.Equal(list.Count, belief);
```

Comment:
Lists are reference types, and `list` is passed down as an argument. This means that the contents of the list cannot be assumed to be unaltered. As such, if this test aims to assert that the list should contain exactly one item, comparing the belief's to the current list's count can give wrong results.

I would store the lists' current count in a variable, before creating the belief, and use that to compare. Or, just keep the `1` but explain why you use `1`.

Reply:
The test does not aim to assert that the list should contain exactly one item. It aims to assert that the belief is modified correctly if the shouldUpdate condition is satisfied. It does exactly that now, no matter what happens to the list.

Reply:
Right now you are testing something which is by definition equal: the length of X should equal to the list in belief Y, where Y is a wrapper around X. You are not testing whether it _changed_, but whether it is equal.